markdown/src/org/intellij/plugins/markdown/structureView/MarkdownStructureElement.java

```diff
@@ -38,107 +42,149 @@ public String getLocationString() {
+      return filePresent!=null ? filePresent : DUMMY_PRESENTATION;
     }
-  }
 
-  @Nullable
-  private ItemPresentation getPresentationImpl() {
-    if (myElement instanceof PsiFileImpl) {
-      return ((PsiFileImpl)myElement).getPresentation();
-    }
-    if (myElement instanceof NavigationItem) {
-      return ((NavigationItem)myElement).getPresentation();
+    if (getElement() instanceof NavigationItem) {
+      final ItemPresentation itemPresent = ((NavigationItem) getElement()).getPresentation();
+      if (itemPresent != null) {
+        return itemPresent;
+      }
     }
 
-    return null;
+    return DUMMY_PRESENTATION;
   }
 
   @NotNull
   @Override
-  public TreeElement[] getChildren() {
-    final PsiElement parentToTraverse = myElement instanceof MarkdownFile ? myElement.getFirstChild() : myElement;
+  public Collection<StructureViewTreeElement> getChildrenBase() {
+    final List<StructureViewTreeElement> childrenElements = new ArrayList<StructureViewTreeElement>();
 
-    if (hasTrivialChild(parentToTraverse)) return EMPTY_ARRAY;
+    final PsiElement myElement = getElement();
+    if(myElement==null) return childrenElements;
 
-    List<TreeElement> result = new ArrayList<>();
-    for (PsiElement element : parentToTraverse.getChildren()) {
-      final IElementType type = element.getNode().getElementType();
-      if (!PRESENTABLE_TYPES.contains(type)) {
-        continue;
+    PsiElement nextSibling = myElement instanceof MarkdownFile ?
+            myElement.getFirstChild().getFirstChild() :
+            myElement.getNextSibling();
+
+    PsiElement maxContentLevel = null;
+
+    while (nextSibling != null) {
+      if (isSameLevelOrHigher(nextSibling, myElement)) {
+        break;
       }
 
-      result.add(new MarkdownStructureElement(element));
+      if (maxContentLevel == null || isSameLevelOrHigher(nextSibling, maxContentLevel)) {
+        maxContentLevel = nextSibling;
+
+        final IElementType type = nextSibling.getNode().getElementType();
+        if (PRESENTABLE_TYPES.contains(type)) {
+          childrenElements.add(new MarkdownStructureElement(nextSibling));
+        }
+      }
+
+      nextSibling = nextSibling.getNextSibling();
     }
-    return ArrayUtil.toObjectArray(result, TreeElement.class);
+
+    return childrenElements;
   }
 
-  public static boolean hasTrivialChild(@NotNull PsiElement parentToTraverse) {
-    if ((parentToTraverse instanceof MarkdownListItemImpl
-         || parentToTraverse instanceof MarkdownBlockQuoteImpl) &&
-        ((MarkdownCompositePsiElementBase)parentToTraverse).hasTrivialChildren()) {
-      return true;
+
+  private boolean isSameLevelOrHigher(@NotNull PsiElement psiA, @NotNull PsiElement psiB) {
+    IElementType typeA = psiA.getNode().getElementType();
+    IElementType typeB = psiB.getNode().getElementType();
+
+    return headerLevel(typeA) <= headerLevel(typeB);
+  }
+
+
+  private int headerLevel(@NotNull IElementType curLevelType) {
```

Comment:
One may get header level from the `org.intellij.plugins.markdown.lang.psi.impl.MarkdownHeaderImpl#getHeaderNumber` (needs visibility escalating)